book/styles.md

```diff
@@ -560,115 +705,137 @@ def style(node, rules):
-    together, if you used the scores suggested above.
--   Tags, class, and identifiers are not the only selectors! Another
-    commonly-used selector is the *descendent* selector; the syntax is
-    multiple space-separated selectors, like `ul strong`, which
-    selects all `<strong>` elements^[This is basically a `<b>` tag but
-    with a hipper name. The idea was to drop all visual aspects from
-    element names.] with a `<ul>` ancestor. Implement descendent
-    selectors, both in parsing and with a new `DescendentSelector`
-    class. Scoring for descendent selectors works just like in
-    `AndSelector`. Make sure that something like `section .warning`
-    selects warnings inside sections, while `section.warning` selects
-    warnings that *are* sections.
+*Shortcuts*: CSS "shortcut properties" set multiple related CSS
+properties at the same time; `margin`, `padding`, `border-width`, and
+`font` are all popular shortcuts. Implement these four shortcut
+properties. If you do it in the `body` parsing function, you won't
+need to change the rest of the code. Start with the case where all of
+the subproperties are specified, then add default values.
+
+*Comma*: CSS allows a rule to have multiple, comma-separated
+selectors, which is basically the same as multiple rules with the same
+body. Implement this and use it to shorten and shorten the browser
+style sheet.
+
+*Width/Height*: Add support to block layout objects for the `width`
+and `height` properties. These can either be a pixel value, which
+directly sets the width or height of the layout object, or the word
+`auto`, in which case the existing layout algorithm is used.
+
+*Percentages*: Most places where you can specify a pixel value in CSS,
+you can also write a percentage value like `50%`. When you do that for
+`margin`, `border`, or `padding` properties, it's relative to the
+layout object's width, while when you do it for `font-weight` it's
+relative to the parent's font size. Implement percentage values for
+all of these properties.
+
+*Combinations*: Sometimes you want to select an element by tag *and*
+class. You do this by concatenating the selectors without anything in
+between: `span.announce` selects elements that match both `span` and
+`.announce`. Implement a new `AndSelector` class to represent these
+and modify the parser to parse them. Sum priorities.[^lexicographic]
+
+[^lexicographic]: You're supposed to use lexicographic scoring for
+    these `AndSelector` things, but sums will work fine as long as no
+    one strings more than 16 selectors together.
+
+*Descendants*: When multiple selectors are separated with spaces, like
+`ul b`, that selects all `<b>` elements with a `<ul>` ancestor.
+Implement descendent selectors; scoring for descendent selectors works
+just like for comnbination selectors. Make sure that something like `section
+.warning` selects warnings inside sections, while `section.warning`
+selects warnings that *are* sections.
```

Comment:
Another fun exercise:

*Ancestor selectors*: There is no way in CSS at present to do the inverse of descendant selectors. For example, there is no way to have a CSS rule that matches "all ul tags that contain a b in the subtree". Can you guess why? Try to implement ancestor selectors and analyze the speed of your algorithm.